File: src/frontend/ged/lib/app/data/model/ged_documento_detalhe_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:ged/app/infra/infra_imports.dart';
import 'package:ged/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';
import 'package:ged/app/data/domain/domain_imports.dart';

class GedDocumentoDetalheModel {
	int? id;
	int? idGedDocumentoCabecalho;
	int? idGedTipoDocumento;
	String? nome;
	String? descricao;
	String? palavrasChave;
	String? podeExcluir;
	String? podeAlterar;
	String? assinado;
	DateTime? dataFimVigencia;
	DateTime? dataExclusao;
	GedTipoDocumentoModel? gedTipoDocumentoModel;

	GedDocumentoDetalheModel({
		this.id,
		this.idGedDocumentoCabecalho,
		this.idGedTipoDocumento,
		this.nome,
		this.descricao,
		this.palavrasChave,
		this.podeExcluir,
		this.podeAlterar,
		this.assinado,
		this.dataFimVigencia,
		this.dataExclusao,
		this.gedTipoDocumentoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'nome',
		'descricao',
		'palavras_chave',
		'pode_excluir',
		'pode_alterar',
		'assinado',
		'data_fim_vigencia',
		'data_exclusao',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Nome',
		'Descricao',
		'Palavras Chave',
		'Pode Excluir',
		'Pode Alterar',
		'Assinado',
		'Data Fim Vigencia',
		'Data Exclusao',
	];

	GedDocumentoDetalheModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idGedDocumentoCabecalho = jsonData['idGedDocumentoCabecalho'];
		idGedTipoDocumento = jsonData['idGedTipoDocumento'];
		nome = jsonData['nome'];
		descricao = jsonData['descricao'];
		palavrasChave = jsonData['palavrasChave'];
		podeExcluir = GedDocumentoDetalheDomain.getPodeExcluir(jsonData['podeExcluir']);
		podeAlterar = GedDocumentoDetalheDomain.getPodeAlterar(jsonData['podeAlterar']);
		assinado = GedDocumentoDetalheDomain.getAssinado(jsonData['assinado']);
		dataFimVigencia = jsonData['dataFimVigencia'] != null ? DateTime.tryParse(jsonData['dataFimVigencia']) : null;
		dataExclusao = jsonData['dataExclusao'] != null ? DateTime.tryParse(jsonData['dataExclusao']) : null;
		gedTipoDocumentoModel = jsonData['gedTipoDocumentoModel'] == null ? GedTipoDocumentoModel() : GedTipoDocumentoModel.fromJson(jsonData['gedTipoDocumentoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idGedDocumentoCabecalho'] = idGedDocumentoCabecalho != 0 ? idGedDocumentoCabecalho : null;
		jsonData['idGedTipoDocumento'] = idGedTipoDocumento != 0 ? idGedTipoDocumento : null;
		jsonData['nome'] = nome;
		jsonData['descricao'] = descricao;
		jsonData['palavrasChave'] = palavrasChave;
		jsonData['podeExcluir'] = GedDocumentoDetalheDomain.setPodeExcluir(podeExcluir);
		jsonData['podeAlterar'] = GedDocumentoDetalheDomain.setPodeAlterar(podeAlterar);
		jsonData['assinado'] = GedDocumentoDetalheDomain.setAssinado(assinado);
		jsonData['dataFimVigencia'] = dataFimVigencia != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataFimVigencia!) : null;
		jsonData['dataExclusao'] = dataExclusao != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataExclusao!) : null;
		jsonData['gedTipoDocumentoModel'] = gedTipoDocumentoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idGedDocumentoCabecalho = plutoRow.cells['idGedDocumentoCabecalho']?.value;
		idGedTipoDocumento = plutoRow.cells['idGedTipoDocumento']?.value;
		nome = plutoRow.cells['nome']?.value;
		descricao = plutoRow.cells['descricao']?.value;
		palavrasChave = plutoRow.cells['palavrasChave']?.value;
		podeExcluir = plutoRow.cells['podeExcluir']?.value != '' ? plutoRow.cells['podeExcluir']?.value : 'S';
		podeAlterar = plutoRow.cells['podeAlterar']?.value != '' ? plutoRow.cells['podeAlterar']?.value : 'S';
		assinado = plutoRow.cells['assinado']?.value != '' ? plutoRow.cells['assinado']?.value : 'S';
		dataFimVigencia = Util.stringToDate(plutoRow.cells['dataFimVigencia']?.value);
		dataExclusao = Util.stringToDate(plutoRow.cells['dataExclusao']?.value);
		gedTipoDocumentoModel = GedTipoDocumentoModel();
		gedTipoDocumentoModel?.nome = plutoRow.cells['gedTipoDocumentoModel']?.value;
	}	

	GedDocumentoDetalheModel clone() {
		return GedDocumentoDetalheModel(
			id: id,
			idGedDocumentoCabecalho: idGedDocumentoCabecalho,
			idGedTipoDocumento: idGedTipoDocumento,
			nome: nome,
			descricao: descricao,
			palavrasChave: palavrasChave,
			podeExcluir: podeExcluir,
			podeAlterar: podeAlterar,
			assinado: assinado,
			dataFimVigencia: dataFimVigencia,
			dataExclusao: dataExclusao,
		);			
	}

	
}

File: backend/ged/src/model/ged_documento_detalhe_model.py
from src import db
from src.model.ged_tipo_documento_model import GedTipoDocumentoModel


class GedDocumentoDetalheModel(db.Model):
    __tablename__ = 'ged_documento_detalhe'

    id = db.Column(db.Integer, primary_key=True)
    nome = db.Column(db.String(100))
    descricao = db.Column(db.Text)
    palavras_chave = db.Column(db.String(250))
    pode_excluir = db.Column(db.String(1))
    pode_alterar = db.Column(db.String(1))
    assinado = db.Column(db.String(1))
    data_fim_vigencia = db.Column(db.DateTime)
    data_exclusao = db.Column(db.DateTime)
    id_ged_tipo_documento = db.Column(db.Integer, db.ForeignKey('ged_tipo_documento.id'))
    id_ged_documento_cabecalho = db.Column(db.Integer, db.ForeignKey('ged_documento_cabecalho.id'))

    ged_tipo_documento_model = db.relationship('GedTipoDocumentoModel', foreign_keys=[id_ged_tipo_documento])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_ged_documento_cabecalho = data.get('idGedDocumentoCabecalho')
        self.id_ged_tipo_documento = data.get('idGedTipoDocumento')
        self.nome = data.get('nome')
        self.descricao = data.get('descricao')
        self.palavras_chave = data.get('palavrasChave')
        self.pode_excluir = data.get('podeExcluir')
        self.pode_alterar = data.get('podeAlterar')
        self.assinado = data.get('assinado')
        self.data_fim_vigencia = data.get('dataFimVigencia')
        self.data_exclusao = data.get('dataExclusao')

    def serialize(self):
        return {
            'id': self.id,
            'idGedDocumentoCabecalho': self.id_ged_documento_cabecalho,
            'idGedTipoDocumento': self.id_ged_tipo_documento,
            'nome': self.nome,
            'descricao': self.descricao,
            'palavrasChave': self.palavras_chave,
            'podeExcluir': self.pode_excluir,
            'podeAlterar': self.pode_alterar,
            'assinado': self.assinado,
            'dataFimVigencia': self.data_fim_vigencia.isoformat(),
            'dataExclusao': self.data_exclusao.isoformat(),
            'gedTipoDocumentoModel': self.ged_tipo_documento_model.serialize() if self.ged_tipo_documento_model else None,
        }